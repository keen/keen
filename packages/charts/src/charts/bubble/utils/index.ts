import { randomizeVector } from './animation.utils';
import { generateBubbles, bubbleColorScale } from './chart.utils';

export { randomizeVector, generateBubbles, bubbleColorScale };
