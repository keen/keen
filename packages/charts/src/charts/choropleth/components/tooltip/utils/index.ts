import calculatePartialPercents from './calculate-partials-percents';

export { calculatePartialPercents };
