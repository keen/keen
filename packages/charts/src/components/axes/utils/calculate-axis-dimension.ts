import { select } from 'd3-selection';

import getRotatedDimension from './get-rotated-dimension';
import createElementIdSelector from './create-element-id-selector';

import { Axis, Orientation } from '../../../types';

type Options = {
  /** Root visualization element */
  svgElement: SVGElement;
  /** Axis theme settings */
  axisTheme: Axis;
  /** Axis orientation */
  orientation: Orientation;
  /** Axis title  */
  axisTitle?: string;
};

/**
 * Calculates computed dimension of all axis elements.
 *
 * @param svgElement - root svg element
 * @param axisTheme - axis theming
 * @param orientation - axis orientation
 * @param axisTitle - axis title
 * @return width and height of axis elements group
 *
 */
const calculateAxisDimension = ({
  axisTheme,
  axisTitle,
  svgElement,
  orientation,
}: Options): Partial<DOMRect> => {
  const {
    enabled: axisEnabled,
    padding,
    tickSize,
    tickPadding,
    title: { padding: titlePadding },
    labels: { enabled: labelsEnabled, radiusAngle },
  } = axisTheme;

  const rootContainer = select(svgElement);

  const labelsSelector = createElementIdSelector(`${orientation}-ruler-label`);
  const titleSelector = createElementIdSelector(`${orientation}-axis-title`);

  let maxLabel = '';
  let longestNode: SVGTextElement = null;

  rootContainer
    .selectAll(labelsSelector)
    .nodes()
    .forEach((node) => {
      const labelElement = select(node);
      const text = labelElement.text();

      if (text.length > maxLabel.length) {
        maxLabel = text;
        longestNode = node as SVGTextElement;
      }
    });

  const textDimension = longestNode
    ? longestNode.getBBox()
    : {
        width: 0,
        height: 0,
      };

  let dimension = { width: 0, height: 0 };

  if (axisEnabled && labelsEnabled) {
    dimension = getRotatedDimension({ radiusAngle, dimension: textDimension });
  }

  if (axisTitle) {
    const titleElement = rootContainer.select(titleSelector);
    const titleNode = titleElement && titleElement.node();
    const titleBbox = titleNode
      ? (titleNode as SVGTextElement).getBBox()
      : {
          width: 0,
          height: 0,
        };

    if (orientation === Orientation.VERTICAL) {
      dimension = {
        height: dimension.height,
        width: dimension.width + titleBbox.height + titlePadding,
      };
    } else {
      dimension = {
        width: dimension.width,
        height: dimension.height + titleBbox.height + titlePadding,
      };
    }
  }

  if (axisEnabled) {
    const { width, height } = dimension;
    if (orientation === Orientation.VERTICAL) {
      dimension = {
        width: width + tickSize + tickPadding + padding,
        height,
      };
    } else {
      dimension = {
        width: width,
        height:
          height +
          Math.abs(tickSize + tickPadding - textDimension.height) +
          padding,
      };
    }
  }

  return dimension;
};

export default calculateAxisDimension;
