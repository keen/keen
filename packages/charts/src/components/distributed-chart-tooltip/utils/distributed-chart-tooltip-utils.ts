import { DataSelector, GroupMode } from '../../../types';
import {
  formatValue as valueFormatter,
  getFromPath,
  TooltipFormatter,
} from '@keen.io/charts-utils';

/**
 * Prepare selectors for bar while mouse hover
 *
 * @param isPercentage - is stackMode 'percent' on
 * @param selector - selector of active bar
 * @param data - data series
 * @param percentageData - calculated percentage data
 * @param formatValue - function for tooltip value format
 * @return a value that will be displayed in tooltip
 *
 */
export const getLabel = ({
  isPercentage,
  selector,
  data,
  percentageData,
  formatValue,
}: {
  selector: DataSelector;
  isPercentage: boolean;
  data: Record<string, any>[];
  percentageData: Record<string, any>[];
  formatValue?: TooltipFormatter;
}) => {
  const value = valueFormatter(getFromPath(data, selector), formatValue);
  if (isPercentage) {
    return {
      value: `${getFromPath(percentageData, selector).toFixed(1)}%`,
      change: `(${value})`,
    };
  }

  return { value };
};

/**
 * Prepare selectors for bar while mouse hover
 *
 * @param groupMode - groupMode option
 * @param keys - keys for the calculations
 * @param disabledKeys - keys disabled for calculation/display
 * @param colors - colors array
 * @param selector - color of the bar
 * @return return an array of DataSelector objects with a color
 *
 */

export const getSelectors = ({
  groupMode,
  keys,
  disabledKeys,
  colors,
  selector,
}: {
  groupMode: GroupMode;
  disabledKeys: string[];
  keys: string[];
  colors: string[];
  selector: { selector: DataSelector; color: string };
}): { selector: DataSelector; color: string }[] => {
  if (groupMode === 'stacked') {
    const { selector: dataSelector } = selector;
    const [index] = dataSelector;
    const selectors: { selector: DataSelector; color: string }[] = [];

    keys.forEach((key: string, idx: number) => {
      if (!disabledKeys.includes(key)) {
        selectors.push({
          selector: [index, key],
          color: colors[idx],
        });
      }
    });

    return selectors;
  }

  return [selector];
};
