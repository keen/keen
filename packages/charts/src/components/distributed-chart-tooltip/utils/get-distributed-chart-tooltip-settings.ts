import {
  formatValue as valueFormatter,
  getFromPath,
  getKeysDifference,
  ScaleSettings,
  TooltipFormatter,
  transformToPercent,
} from '@keen.io/charts-utils';

import { DataSelector } from '../../../types';
import { getLabel } from '../index';

type TooltipSettings = {
  data: Record<string, any>[];
  keys: string[];
  disabledKeys: string[];
  selectors: { selector: DataSelector; color: string }[];
  isPercentage: boolean;
  labelSelector: string;
  formatValue?: TooltipFormatter;
  scaleSettings?: ScaleSettings;
};

const getDistributedChartTooltipSettings = ({
  data,
  keys,
  disabledKeys,
  selectors,
  isPercentage,
  labelSelector,
  formatValue,
  scaleSettings,
}: TooltipSettings) => {
  const { precision, formatLabel } = scaleSettings;

  const percentageData = isPercentage
    ? transformToPercent(data, getKeysDifference(keys, disabledKeys))
    : [];

  const [firstSelector] = selectors;
  const [index] = firstSelector.selector;

  const isGrouped = data.length > 1;

  const getTooltipLabel = () => {
    if (typeof index === 'number' && (precision || isGrouped)) {
      return `${valueFormatter(data[index][labelSelector], formatLabel)}`;
    }
    return null;
  };

  const totalValue =
    selectors.length > 1
      ? keys.reduce((acc: number, keyName: string) => {
          if (typeof index !== 'number') return acc;
          const value = data[index][keyName];
          if (typeof value !== 'number') return acc;
          return acc + value;
        }, 0)
      : null;

  const percentValue = isPercentage
    ? selectors.reduce((acc, { selector }) => {
        return (
          acc + parseFloat(getFromPath(percentageData, selector).toFixed(2))
        );
      }, 0)
    : null;

  const items = selectors.map(({ color, selector }) => {
    return {
      color,
      data: {
        label: `${selector[1]}`,
        ...getLabel({
          data,
          selector,
          percentageData,
          isPercentage,
          formatValue,
        }),
      },
    };
  });

  return {
    getTooltipLabel,
    percentValue,
    totalValue,
    items,
  };
};

export default getDistributedChartTooltipSettings;
