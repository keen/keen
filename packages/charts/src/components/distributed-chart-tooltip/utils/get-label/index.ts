import getLabel from './get-label';

export default getLabel;
