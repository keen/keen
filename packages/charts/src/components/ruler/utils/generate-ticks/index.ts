import generateTicks from './generate-ticks';

export default generateTicks;
