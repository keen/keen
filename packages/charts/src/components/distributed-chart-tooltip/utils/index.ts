import getDistributedChartTooltipSettings from './get-distributed-chart-tooltip-settings';
import { getLabel, getSelectors } from './distributed-chart-tooltip-utils';

export { getDistributedChartTooltipSettings, getSelectors, getLabel };
