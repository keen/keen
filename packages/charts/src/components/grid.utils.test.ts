import { scaleBand, scaleLinear } from 'd3-scale';

import { generateGridLines, AxisType } from './grid.utils';

describe('@keen.io/charts', () => {
  describe('<Grid /> - utils', () => {
    describe('generateGridLines()', () => {
      const margins = { top: 10, right: 10, bottom: 10, left: 10 };
      const dimension = {
        width: 100,
        height: 100,
      };

      it('should calculate lines coordinates for band scale x', () => {
        const domain = ['January', 'February'];
        const scale = scaleBand().range([0, 100]).domain(domain);

        const lines = generateGridLines({
          scale,
          dimension,
          margins,
          axisType: AxisType.X,
        });

        expect(lines).toMatchSnapshot();
        expect(lines.length).toEqual(4);
      });

      it('should calculate lines coordinates for band scale y', () => {
        const domain = ['January', 'February', 'March', 'April'];
        const scale = scaleBand().range([0, 200]).domain(domain);

        const lines = generateGridLines({
          scale,
          dimension,
          margins,
          axisType: AxisType.Y,
        });

        expect(lines).toMatchSnapshot();
        expect(lines.length).toEqual(6);
      });

      it('should calculate lines coordinates for linear scale y', () => {
        const scale = scaleLinear().range([0, 100]).domain([10, 60]);

        const lines = generateGridLines({
          scale,
          dimension,
          margins,
          axisType: AxisType.Y,
        });

        expect(scale.ticks().length).toEqual(lines.length);
        expect(lines).toMatchSnapshot();
      });

      it('should calculate lines coordinates for linear scale x', () => {
        const scale = scaleLinear().range([0, 100]).domain([10, 90]);

        const lines = generateGridLines({
          scale,
          dimension,
          margins,
          axisType: AxisType.X,
        });

        expect(scale.ticks().length).toEqual(lines.length);
        expect(lines).toMatchSnapshot();
      });
    });
  });
});
