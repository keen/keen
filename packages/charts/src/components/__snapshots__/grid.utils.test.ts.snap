// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@keen.io/charts <Grid /> - utils generateGridLines() should calculate lines coordinates for band scale x 1`] = `
Array [
  Object {
    "stroke": undefined,
    "x1": 25,
    "x2": 25,
    "y1": 10,
    "y2": 90,
  },
  Object {
    "stroke": undefined,
    "x1": 75,
    "x2": 75,
    "y1": 10,
    "y2": 90,
  },
  Object {
    "x1": 10,
    "x2": 10,
    "y1": 9.5,
    "y2": 90.5,
  },
  Object {
    "x1": 90,
    "x2": 90,
    "y1": 9.5,
    "y2": 90.5,
  },
]
`;

exports[`@keen.io/charts <Grid /> - utils generateGridLines() should calculate lines coordinates for band scale y 1`] = `
Array [
  Object {
    "stroke": undefined,
    "x1": 10,
    "x2": 90,
    "y1": 25,
    "y2": 25,
  },
  Object {
    "stroke": undefined,
    "x1": 10,
    "x2": 90,
    "y1": 75,
    "y2": 75,
  },
  Object {
    "stroke": undefined,
    "x1": 10,
    "x2": 90,
    "y1": 125,
    "y2": 125,
  },
  Object {
    "stroke": undefined,
    "x1": 10,
    "x2": 90,
    "y1": 175,
    "y2": 175,
  },
  Object {
    "x1": 9.5,
    "x2": 90.5,
    "y1": 10,
    "y2": 10,
  },
  Object {
    "x1": 9.5,
    "x2": 90.5,
    "y1": 90,
    "y2": 90,
  },
]
`;

exports[`@keen.io/charts <Grid /> - utils generateGridLines() should calculate lines coordinates for linear scale x 1`] = `
Array [
  Object {
    "stroke": undefined,
    "x1": 0,
    "x2": 0,
    "y1": 10,
    "y2": 90,
  },
  Object {
    "stroke": undefined,
    "x1": 12.5,
    "x2": 12.5,
    "y1": 10,
    "y2": 90,
  },
  Object {
    "stroke": undefined,
    "x1": 25,
    "x2": 25,
    "y1": 10,
    "y2": 90,
  },
  Object {
    "stroke": undefined,
    "x1": 37.5,
    "x2": 37.5,
    "y1": 10,
    "y2": 90,
  },
  Object {
    "stroke": undefined,
    "x1": 50,
    "x2": 50,
    "y1": 10,
    "y2": 90,
  },
  Object {
    "stroke": undefined,
    "x1": 62.5,
    "x2": 62.5,
    "y1": 10,
    "y2": 90,
  },
  Object {
    "stroke": undefined,
    "x1": 75,
    "x2": 75,
    "y1": 10,
    "y2": 90,
  },
  Object {
    "stroke": undefined,
    "x1": 87.5,
    "x2": 87.5,
    "y1": 10,
    "y2": 90,
  },
  Object {
    "stroke": undefined,
    "x1": 100,
    "x2": 100,
    "y1": 10,
    "y2": 90,
  },
]
`;

exports[`@keen.io/charts <Grid /> - utils generateGridLines() should calculate lines coordinates for linear scale y 1`] = `
Array [
  Object {
    "stroke": undefined,
    "x1": 10,
    "x2": 90,
    "y1": 0,
    "y2": 0,
  },
  Object {
    "stroke": undefined,
    "x1": 10,
    "x2": 90,
    "y1": 10,
    "y2": 10,
  },
  Object {
    "stroke": undefined,
    "x1": 10,
    "x2": 90,
    "y1": 20,
    "y2": 20,
  },
  Object {
    "stroke": undefined,
    "x1": 10,
    "x2": 90,
    "y1": 30,
    "y2": 30,
  },
  Object {
    "stroke": undefined,
    "x1": 10,
    "x2": 90,
    "y1": 40,
    "y2": 40,
  },
  Object {
    "stroke": undefined,
    "x1": 10,
    "x2": 90,
    "y1": 50,
    "y2": 50,
  },
  Object {
    "stroke": undefined,
    "x1": 10,
    "x2": 90,
    "y1": 60,
    "y2": 60,
  },
  Object {
    "stroke": undefined,
    "x1": 10,
    "x2": 90,
    "y1": 70,
    "y2": 70,
  },
  Object {
    "stroke": undefined,
    "x1": 10,
    "x2": 90,
    "y1": 80,
    "y2": 80,
  },
  Object {
    "stroke": undefined,
    "x1": 10,
    "x2": 90,
    "y1": 90,
    "y2": 90,
  },
  Object {
    "stroke": undefined,
    "x1": 10,
    "x2": 90,
    "y1": 100,
    "y2": 100,
  },
]
`;
